File: examples/policies/typescript/test-pattern.ts
/**
 * AxonFlow Policy Management - Test Pattern
 *
 * This example demonstrates how to test regex patterns
 * before creating policies. This helps ensure your patterns
 * work correctly and catch the right inputs.
 *
 * VALIDATION: This example exits with code 1 if any assertion fails.
 */

import { AxonFlow } from '@axonflow/sdk';

const failures: string[] = [];

function assertCheck(condition: boolean, message: string): void {
  if (condition) {
    console.log(`   ✓ PASS: ${message}`);
  } else {
    console.log(`   ❌ FAIL: ${message}`);
    failures.push(message);
  }
}

async function main() {
  const client = new AxonFlow({
    endpoint: process.env.AXONFLOW_ENDPOINT || 'http://localhost:8080',
  });

  console.log('AxonFlow Policy Management - Pattern Testing');
  console.log('='.repeat(60));

  try {
    // 1. Test a credit card pattern
    console.log('\n1. Testing credit card pattern...');

    const ccPattern = '\\b(?:\\d{4}[- ]?){3}\\d{4}\\b';
    const ccTestInputs = [
      '4111-1111-1111-1111',      // Valid Visa format with dashes
      '4111111111111111',          // Valid Visa format no dashes
      '4111 1111 1111 1111',       // Valid with spaces
      'not-a-card',                // Invalid
      '411111111111111',           // Too short (15 digits)
      '41111111111111111',         // Too long (17 digits)
      'My card is 5500-0000-0000-0004',  // Embedded in text
    ];

    const ccResult = await client.testPattern(ccPattern, ccTestInputs);

    console.log(`   Pattern: ${ccPattern}`);
    console.log(`   Valid regex: ${ccResult.valid}`);
    console.log('\n   Results:');

    ccResult.matches.forEach((match) => {
      const icon = match.matched ? '\u2713 MATCH' : '\u2717 no match';
      console.log(`   ${icon}  "${match.input}"`);
    });

    assertCheck(ccResult.valid === true, 'Credit card pattern is valid regex');
    assertCheck(ccResult.matches.length === 7, 'All 7 credit card test inputs were evaluated');
    // Valid formats should match
    assertCheck(ccResult.matches[0].matched === true, '4111-1111-1111-1111 (dashes) matches');
    assertCheck(ccResult.matches[1].matched === true, '4111111111111111 (no dashes) matches');
    assertCheck(ccResult.matches[2].matched === true, '4111 1111 1111 1111 (spaces) matches');
    // Invalid formats should not match
    assertCheck(ccResult.matches[3].matched === false, 'not-a-card does not match');
    assertCheck(ccResult.matches[4].matched === false, '15-digit number does not match');
    assertCheck(ccResult.matches[5].matched === false, '17-digit number does not match');
    // Embedded card should match
    assertCheck(ccResult.matches[6].matched === true, 'Embedded card in text matches');

    // 2. Test a US SSN pattern
    console.log('\n2. Testing US SSN pattern...');

    const ssnPattern = '\\b\\d{3}-\\d{2}-\\d{4}\\b';
    const ssnTestInputs = [
      '123-45-6789',               // Valid SSN format
      '000-00-0000',               // Valid format (but invalid SSN)
      'SSN: 987-65-4321',          // Embedded in text
      '123456789',                 // No dashes
      '12-345-6789',               // Wrong grouping
    ];

    const ssnResult = await client.testPattern(ssnPattern, ssnTestInputs);

    console.log(`   Pattern: ${ssnPattern}`);
    console.log('\n   Results:');

    ssnResult.matches.forEach((match) => {
      const icon = match.matched ? '\u2713 MATCH' : '\u2717 no match';
      console.log(`   ${icon}  "${match.input}"`);
    });

    assertCheck(ssnResult.valid === true, 'SSN pattern is valid regex');
    assertCheck(ssnResult.matches.length === 5, 'All 5 SSN test inputs were evaluated');
    assertCheck(ssnResult.matches[0].matched === true, '123-45-6789 matches SSN pattern');
    assertCheck(ssnResult.matches[1].matched === true, '000-00-0000 matches SSN format');
    assertCheck(ssnResult.matches[2].matched === true, 'Embedded SSN in text matches');
    assertCheck(ssnResult.matches[3].matched === false, 'No-dashes SSN does not match');
    assertCheck(ssnResult.matches[4].matched === false, 'Wrong grouping does not match');

    // 3. Test an email pattern
    console.log('\n3. Testing email pattern...');

    const emailPattern = '[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}';
    const emailTestInputs = [
      'user@example.com',
      'first.last@company.org',
      'test+filter@gmail.com',
      'invalid-email',
      '@missing-local.com',
      'no-domain@',
    ];

    const emailResult = await client.testPattern(emailPattern, emailTestInputs);

    console.log(`   Pattern: ${emailPattern}`);
    console.log('\n   Results:');

    emailResult.matches.forEach((match) => {
      const icon = match.matched ? '\u2713 MATCH' : '\u2717 no match';
      console.log(`   ${icon}  "${match.input}"`);
    });

    assertCheck(emailResult.valid === true, 'Email pattern is valid regex');
    assertCheck(emailResult.matches.length === 6, 'All 6 email test inputs were evaluated');
    assertCheck(emailResult.matches[0].matched === true, 'user@example.com matches');
    assertCheck(emailResult.matches[1].matched === true, 'first.last@company.org matches');
    assertCheck(emailResult.matches[2].matched === true, 'test+filter@gmail.com matches');
    assertCheck(emailResult.matches[3].matched === false, 'invalid-email does not match');
    assertCheck(emailResult.matches[4].matched === false, '@missing-local.com does not match');
    assertCheck(emailResult.matches[5].matched === false, 'no-domain@ does not match');

    // 4. Test SQL injection pattern
    console.log('\n4. Testing SQL injection pattern...');

    const sqliPattern = '(?i)\\b(union\\s+select|select\\s+.*\\s+from|insert\\s+into|delete\\s+from|drop\\s+table)\\b';
    const sqliTestInputs = [
      'SELECT * FROM users',
      'UNION SELECT password FROM admin',
      'DROP TABLE customers',
      'Normal user query',
      'My name is Robert',
      'INSERT INTO logs VALUES',
    ];

    const sqliResult = await client.testPattern(sqliPattern, sqliTestInputs);

    console.log(`   Pattern: ${sqliPattern.slice(0, 50)}...`);
    console.log('\n   Results:');

    sqliResult.matches.forEach((match) => {
      const icon = match.matched ? '\u2713 BLOCKED' : '\u2717 allowed';
      console.log(`   ${icon}  "${match.input}"`);
    });

    assertCheck(sqliResult.valid === true, 'SQL injection pattern is valid regex');
    assertCheck(sqliResult.matches.length === 6, 'All 6 SQLi test inputs were evaluated');
    // SQL injection patterns should be detected
    assertCheck(sqliResult.matches[0].matched === true, 'SELECT * FROM detected as SQLi');
    assertCheck(sqliResult.matches[1].matched === true, 'UNION SELECT detected as SQLi');
    assertCheck(sqliResult.matches[2].matched === true, 'DROP TABLE detected as SQLi');
    // Normal inputs should not be blocked
    assertCheck(sqliResult.matches[3].matched === false, 'Normal query is allowed');
    assertCheck(sqliResult.matches[4].matched === false, 'Regular name is allowed');
    assertCheck(sqliResult.matches[5].matched === true, 'INSERT INTO detected as SQLi');

    // 5. Test an invalid pattern
    console.log('\n5. Testing invalid pattern (error handling)...');

    let invalidHandledCorrectly = false;
    try {
      const invalidPattern = '([unclosed';
      const invalidResult = await client.testPattern(invalidPattern, ['test']);

      if (!invalidResult.valid) {
        console.log(`   Pattern: ${invalidPattern}`);
        console.log(`   Valid: false`);
        console.log(`   Error: ${invalidResult.error}`);
        invalidHandledCorrectly = true;
        assertCheck(invalidResult.valid === false, 'Invalid pattern returns valid=false');
        assertCheck(typeof invalidResult.error === 'string', 'Invalid pattern returns error message');
      }
    } catch (e) {
      console.log('   Server rejected invalid pattern (expected)');
      invalidHandledCorrectly = true;
    }
    assertCheck(invalidHandledCorrectly, 'Invalid regex pattern is properly handled');

    // Summary
    console.log('\n' + '='.repeat(60));
    console.log('Pattern Testing Summary');
    console.log('='.repeat(60));
    console.log(`
Best Practices:
  1. Always test patterns before creating policies
  2. Include edge cases in your test inputs
  3. Test with real-world examples from your domain
  4. Consider case sensitivity (use (?i) for case-insensitive)
  5. Use word boundaries (\\b) to avoid partial matches
`);

  } catch (error) {
    if (error instanceof Error) {
      console.error('\nError:', error.message);
    }
    failures.push(`Unexpected error: ${error instanceof Error ? error.message : error}`);
  }

  // Final assertion summary
  console.log('='.repeat(60));
  console.log('Assertion Summary');
  console.log('='.repeat(60));
  if (failures.length === 0) {
    console.log('All assertions passed!');
  } else {
    console.log(`${failures.length} assertion(s) failed:`);
    failures.forEach((f) => console.log(`  - ${f}`));
  }

  process.exit(failures.length > 0 ? 1 : 0);
}

main();

File: examples/policies/python/test_pattern.py
#!/usr/bin/env python3
"""
AxonFlow Policy Management - Test Pattern

VALIDATION: This example exits with code 1 if any assertion fails.

This example demonstrates how to test regex patterns
before creating policies. This helps ensure your patterns
work correctly and catch the right inputs.

Run with: python test_pattern.py
Prerequisites: docker compose up -d
"""

import asyncio
import os
import sys

from axonflow import AxonFlow

failures: list[str] = []


def assert_check(condition: bool, message: str) -> None:
    """Check a condition and record failure if false."""
    if condition:
        print(f"   ✓ PASS: {message}")
    else:
        print(f"   ❌ FAIL: {message}")
        failures.append(message)


async def main() -> int:
    """Test various regex patterns."""
    print("AxonFlow Policy Management - Pattern Testing")
    print("=" * 60)

    client = AxonFlow(
        endpoint=os.getenv("AXONFLOW_ENDPOINT", "http://localhost:8080"),
        client_id="test-org-001",
        client_secret="test-secret",
    )

    try:
        # 1. Test a credit card pattern
        print("\n1. Testing credit card pattern...")

        cc_pattern = r"\b(?:\d{4}[- ]?){3}\d{4}\b"
        cc_test_inputs = [
            "4111-1111-1111-1111",  # Valid Visa format with dashes
            "4111111111111111",  # Valid Visa format no dashes
            "4111 1111 1111 1111",  # Valid with spaces
            "not-a-card",  # Invalid
            "411111111111111",  # Too short (15 digits)
            "41111111111111111",  # Too long (17 digits)
            "My card is 5500-0000-0000-0004",  # Embedded in text
        ]

        cc_result = await client.test_pattern(cc_pattern, cc_test_inputs)

        assert_check(cc_result is not None, "Credit card pattern test returned result")
        assert_check(cc_result.valid, "Credit card pattern is valid regex")
        assert_check(len(cc_result.matches) == len(cc_test_inputs), "All inputs were tested")

        print(f"   Pattern: {cc_pattern}")
        print(f"   Valid regex: {cc_result.valid}")
        print("\n   Results:")

        # Verify expected matches
        expected_cc_matches = [True, True, True, False, False, False, True]
        for i, match in enumerate(cc_result.matches):
            icon = "✓ MATCH" if match.matched else "✗ no match"
            print(f'   {icon}  "{match.input}"')
            if i < len(expected_cc_matches):
                assert_check(
                    match.matched == expected_cc_matches[i],
                    f"CC input '{match.input[:20]}...' matched as expected"
                )

        # 2. Test a US SSN pattern
        print("\n2. Testing US SSN pattern...")

        ssn_pattern = r"\b\d{3}-\d{2}-\d{4}\b"
        ssn_test_inputs = [
            "123-45-6789",  # Valid SSN format
            "000-00-0000",  # Valid format (but invalid SSN)
            "SSN: 987-65-4321",  # Embedded in text
            "123456789",  # No dashes
            "12-345-6789",  # Wrong grouping
        ]

        ssn_result = await client.test_pattern(ssn_pattern, ssn_test_inputs)

        assert_check(ssn_result is not None, "SSN pattern test returned result")
        assert_check(ssn_result.valid, "SSN pattern is valid regex")

        print(f"   Pattern: {ssn_pattern}")
        print("\n   Results:")

        expected_ssn_matches = [True, True, True, False, False]
        for i, match in enumerate(ssn_result.matches):
            icon = "✓ MATCH" if match.matched else "✗ no match"
            print(f'   {icon}  "{match.input}"')
            if i < len(expected_ssn_matches):
                assert_check(
                    match.matched == expected_ssn_matches[i],
                    f"SSN input '{match.input}' matched as expected"
                )

        # 3. Test an email pattern
        print("\n3. Testing email pattern...")

        email_pattern = r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}"
        email_test_inputs = [
            "user@example.com",
            "first.last@company.org",
            "test+filter@gmail.com",
            "invalid-email",
            "@missing-local.com",
            "no-domain@",
        ]

        email_result = await client.test_pattern(email_pattern, email_test_inputs)

        assert_check(email_result is not None, "Email pattern test returned result")
        assert_check(email_result.valid, "Email pattern is valid regex")

        print(f"   Pattern: {email_pattern}")
        print("\n   Results:")

        for match in email_result.matches:
            icon = "✓ MATCH" if match.matched else "✗ no match"
            print(f'   {icon}  "{match.input}"')

        # 4. Test SQL injection pattern
        print("\n4. Testing SQL injection pattern...")

        sqli_pattern = r"(?i)\b(union\s+select|select\s+.*\s+from|insert\s+into|delete\s+from|drop\s+table)\b"
        sqli_test_inputs = [
            "SELECT * FROM users",
            "UNION SELECT password FROM admin",
            "DROP TABLE customers",
            "Normal user query",
            "My name is Robert",
            "INSERT INTO logs VALUES",
        ]

        sqli_result = await client.test_pattern(sqli_pattern, sqli_test_inputs)

        assert_check(sqli_result is not None, "SQLi pattern test returned result")
        assert_check(sqli_result.valid, "SQLi pattern is valid regex")

        print(f"   Pattern: {sqli_pattern[:50]}...")
        print("\n   Results:")

        expected_sqli_matches = [True, True, True, False, False, True]
        for i, match in enumerate(sqli_result.matches):
            icon = "✓ BLOCKED" if match.matched else "✗ allowed"
            print(f'   {icon}  "{match.input}"')
            if i < len(expected_sqli_matches):
                assert_check(
                    match.matched == expected_sqli_matches[i],
                    f"SQLi input '{match.input[:20]}' matched as expected"
                )

        # 5. Test an invalid pattern
        print("\n5. Testing invalid pattern (error handling)...")

        try:
            invalid_pattern = "([unclosed"
            invalid_result = await client.test_pattern(invalid_pattern, ["test"])

            if not invalid_result.valid:
                assert_check(True, "Invalid pattern correctly marked as invalid")
                print(f"   Pattern: {invalid_pattern}")
                print("   Valid: false")
                print(f"   Error: {invalid_result.error}")
            else:
                assert_check(False, "Invalid pattern should be marked as invalid")
        except Exception:
            assert_check(True, "Server rejected invalid pattern (expected)")
            print("   Server rejected invalid pattern (expected)")

    except Exception as e:
        print(f"\nError: {e}")
        failures.append(f"Pattern testing failed: {e}")

    print("\n" + "=" * 60)
    if not failures:
        print("✓ ALL TESTS PASSED")
        print()
        print("Pattern Testing validated:")
        print("  - Credit card pattern matching")
        print("  - SSN pattern matching")
        print("  - Email pattern matching")
        print("  - SQL injection pattern matching")
        print("  - Invalid pattern handling")
        return 0
    else:
        print(f"❌ {len(failures)} TEST(S) FAILED:")
        for f in failures:
            print(f"   - {f}")
        return 1


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))
